examples/cosmos/ics23/hashes.qnt

```diff
@@ -0,0 +1,144 @@
+/// we represent terms as maps that encode trees. For instance,
+/// the term h([ 1, 2 ] + h([ 3, 4 ])) + [ 5, 6 ] corresponds to the tree:
+///
+///       "h"      [ 5, 6 ]
+///     /     \
+/// [ 1, 2 ]   "h"
+///          /
+///      [ 3, 4 ]
+///
+/// The above tree is represented as a map in Quint:
+///
+/// Map([ 0 ] -> { t: "h", b: [] },
+///     [ 0, 0 ] -> { t: "r", b: [ 1, 2 ] },
+///     [ 0, 1 ] -> { t: "h", b: [] } ),
+///     [ 0, 1, 0 ] -> { t: "r", b: [ 3, 4 ] },
+///     [ 1 ] -> { t: "r", b: [ 5, 6 ] })
+///
+/// Igor Konnov, Informal Systems, 2022-2023
+module hashes {
+  /// A sequence of bytes is simply a list of integers
+  type Bytes_t = List[int]
+
+  /// compare two lists of integers (e.g., bytes) lexicographically
+  pure def lessThan(w1: Bytes_t, w2: Bytes_t): bool = {
+    pure val len1 = length(w1)
+    pure val len2 = length(w2)
+    or {
+      len1 < len2 and w1.indices().forall(i => w1[i] == w2[i]),
+      and {
+        len1 == len2,
+        w1.indices().exists(i => and {
+          w1[i] < w2[i],
+          w1.indices().forall(j => j < i implies w1[i] == w2[j])
+        })
+      }
+    }
+  }
+
+  /// Does the sequence of bytes have the nil value.
+  /// In our spec, it is equivalent to empty.
+  pure def isNil(w: Bytes_t): bool = {
+    w == []
+  }
+
+  /// A tree node that represents a fragment of a term
+  type TERM_NODE_T = {
+    // node type: either "h" (for hash), or "r" (for raw sequence)
+    t: str,
+    // the bytes when t = "r", and [] when t = "h"
+    b: Bytes_t
+  }
+
+  /// A word is a map from a path to the term node.
+  /// The first root's child is [ 0 ], the second root's child is [ 1 ],
+  /// the first child of [ 0 ] is [ 0, 0], etc.
+  type Term_t = Bytes_t -> TERM_NODE_T
+
+  /// Compute term length in bytes,
+  /// assuming that a hash occupies 32 bytes (as SHA256 does)
+  pure def termLen(term: Term_t): int = {
+    // the root's children
+    pure val top =
+      keys(term).filter(p => length(p) == 1).map(p => term.get(p))
+    top.fold(0, (s, t) => if (t.t == "h") s + 32 else s + length(t.b))
+  }
+
+  // Construct the term that encodes a raw sequence of bytes
+  pure def raw(bytes: Bytes_t): Term_t = {
+    Map([ 0 ] -> { t: "r", b: bytes })
+  }
+
+  // Hash a term
+  pure def termHash(term: Term_t): Term_t = {
+    // add "h" on top, which has the key [ 0 ], and attach term to it
+    pure val paths = Set([ 0 ]).union(keys(term).map(p => [ 0 ].concat(p)))
+    paths.mapBy(p =>
+      if (p == [ 0 ]) {
+        { t: "h", b: [] }
+      } else {
+        term.get(p.slice(1, length(p)))
+      }
+    )
+  }
+
+  /// Concatenate two terms. Special attention is paid to the case when the
+  /// both terms are raw sequences, which requires them to be merged.
+  pure def termConcat(left: Term_t, right: Term_t): Term_t = {
+    pure val l = left.get([0])
+    pure val r = right.get([0])
+    pure val isLSeq = size(keys(left)) == 1 and l.t == "r"
```

Comment:
This test appears multiple times, and is prone to getting it wrong.
Introduce a predicate def?